store/stats_pusher.go

```diff
@@ -0,0 +1,243 @@
+// delivery.
+func NewWebsocketStatsPusher(url *url.URL) StatsPusher {
+	return &websocketStatsPusher{
+		url:     *url,
+		send:    make(chan []byte, 100), // TODO: figure out a better buffer (circular FIFO?)
+		boot:    &sync.Mutex{},
+		sleeper: utils.NewBackoffSleeper(),
+	}
+}
+
+// Start starts a write pump over a websocket.
+func (w *websocketStatsPusher) Start() error {
+	w.boot.Lock()
+	defer w.boot.Unlock()
+
+	if w.started {
+		return nil
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	w.cancel = cancel
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go w.connectAndWritePump(ctx, wg)
+	wg.Wait()
+	w.started = true
+	return nil
+}
+
+const (
+	// Time allowed to write a message to the peer.
+	writeWait = 10 * time.Second
+
+	// Time allowed to read the next pong message from the peer.
+	pongWait = 60 * time.Second
+
+	// Send pings to peer with this period. Must be less than pongWait.
+	pingPeriod = (pongWait * 9) / 10
+
+	// Maximum message size allowed from peer.
+	maxMessageSize = 512
+)
+
+// Inspired by https://github.com/gorilla/websocket/blob/master/examples/chat/client.go
+// lexical confinement of done chan allows multiple connectAndWritePump routines
+// to clean up independent of itself by reducing shared state. i.e. a passed done, not w.done.
+func (w *websocketStatsPusher) connectAndWritePump(parentCtx context.Context, wg *sync.WaitGroup) {
+	wg.Done()
+	logger.Info("Connecting to linkstats at ", w.url.String())
+
+	for {
+		select {
+		case <-parentCtx.Done():
+			return
+		case <-time.After(w.sleeper.After()):
+			connectionCtx, cancel := context.WithCancel(parentCtx)
+			defer cancel()
+
+			if err := w.connect(connectionCtx); err != nil {
+				logger.Warn("Failed to connect to linkstats (", w.url.String(), "): ", err)
+				break
+			}
+
+			logger.Info("Connected to linkstats at ", w.url.String())
+			w.sleeper.Reset()
+			go w.readPumpForControlMessages(cancel)
+			w.writePump(connectionCtx)
+		}
+	}
+}
+
+var (
+	newline = []byte{'\n'}
+)
+
+// Inspired by https://github.com/gorilla/websocket/blob/master/examples/chat/client.go#L82
+func (w *websocketStatsPusher) writePump(ctx context.Context) {
+	ticker := time.NewTicker(pingPeriod)
+	defer func() {
+		ticker.Stop()
+		wrapLoggerErrorIf(w.conn.Close()) // exclusive responsibility to close ws conn
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case message, open := <-w.send:
+			if !open { // channel closed
+				wrapLoggerErrorIf(w.conn.WriteMessage(websocket.CloseMessage, []byte{}))
+				return
+			}
+
+			wrapLoggerErrorIf(w.conn.SetWriteDeadline(time.Now().Add(writeWait)))
+			writer, err := w.conn.NextWriter(websocket.TextMessage)
+			if err != nil {
+				logger.Error(err)
+				return
+			}
+			_, err = writer.Write(message)
+			wrapLoggerErrorIf(err)
+
+			// Add queued messages to the current websocket message,
+			// batch sending for efficiency.
+			n := len(w.send)
```

Comment:
It's taken from the gorilla websockets example https://github.com/gorilla/websocket/blob/master/examples/chat/client.go#L104. Thought it might help if we have high throughput with taskruns, but I can remove it.

Reply:
It feels like stuff that could get in the way of future work, I just think this kind of thing should be added in later is all.

Reply:
👍 Agreed. It's been removed.